From dced69899416124ce11ce84879998f4d23293a71 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Thu, 9 Jul 2020 15:14:31 +0700
Subject: [PATCH 236/357] soc: renesas: rcar-sysc: Add R8A774A3 support

Refer upstream commit id <bdde3d3ec934839b3c11689ead467099f1c36c12>
support for R-Car M3-W+ and apply it for RZ/G2M v3.0.

Add support for the power areas in the Renesas RZ/G2M v3.0 (R8A774A3).

RZ/G2M v3.0 (aka R8A774A3) is very similar to RZ/G2M (R8A774A1),
which allows for both SoCs to share a driver:
- RZ/G2M v3.0 lacks the A2VC power area, so its area must be
nullified,
- The existing support for the SYSCEXTMASK register applies to v3.0
and later only.

As RZ/G2M v3.0 uses a different compatible value, differentiate
based on that, instead of on the ES version.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/soc/renesas/Kconfig         |  7 ++++++-
 drivers/soc/renesas/Makefile        |  1 +
 drivers/soc/renesas/r8a774a1-sysc.c | 28 ++++++++++++++++++++++++----
 drivers/soc/renesas/rcar-sysc.c     |  4 ++++
 drivers/soc/renesas/rcar-sysc.h     |  2 ++
 5 files changed, 37 insertions(+), 5 deletions(-)

diff --git a/drivers/soc/renesas/Kconfig b/drivers/soc/renesas/Kconfig
index 4b0f70f..82769ec 100644
--- a/drivers/soc/renesas/Kconfig
+++ b/drivers/soc/renesas/Kconfig
@@ -11,6 +11,7 @@ config SOC_RENESAS
 	select SYSC_R8A7745 if ARCH_R8A7745
 	select SYSC_R8A77470 if ARCH_R8A77470
 	select SYSC_R8A774A1 if ARCH_R8A774A1
+	select SYSC_R8A774A3 if ARCH_R8A774A3
 	select SYSC_R8A774B1 if ARCH_R8A774B1
 	select SYSC_R8A774C0 if ARCH_R8A774C0
 	select SYSC_R8A774E1 if ARCH_R8A774E1
@@ -51,7 +52,11 @@ config SYSC_R8A77470
 	select SYSC_RCAR
 
 config SYSC_R8A774A1
-	bool "RZ/G2M System Controller support" if COMPILE_TEST
+	bool "RZ/G2M v1.3 System Controller support" if COMPILE_TEST
+	select SYSC_RCAR
+
+config SYSC_R8A774A3
+	bool "RZ/G2M v3.0 System Controller support" if COMPILE_TEST
 	select SYSC_RCAR
 
 config SYSC_R8A774B1
diff --git a/drivers/soc/renesas/Makefile b/drivers/soc/renesas/Makefile
index d7cbbea..be9024f 100644
--- a/drivers/soc/renesas/Makefile
+++ b/drivers/soc/renesas/Makefile
@@ -7,6 +7,7 @@ obj-$(CONFIG_SYSC_R8A7743)	+= r8a7743-sysc.o
 obj-$(CONFIG_SYSC_R8A7745)	+= r8a7745-sysc.o
 obj-$(CONFIG_SYSC_R8A77470)	+= r8a77470-sysc.o
 obj-$(CONFIG_SYSC_R8A774A1)	+= r8a774a1-sysc.o
+obj-$(CONFIG_SYSC_R8A774A3)	+= r8a774a1-sysc.o
 obj-$(CONFIG_SYSC_R8A774B1)	+= r8a774b1-sysc.o
 obj-$(CONFIG_SYSC_R8A774C0)	+= r8a774c0-sysc.o
 obj-$(CONFIG_SYSC_R8A774E1)	+= r8a774e1-sysc.o
diff --git a/drivers/soc/renesas/r8a774a1-sysc.c b/drivers/soc/renesas/r8a774a1-sysc.c
index 9db51ff..d0f36f5 100644
--- a/drivers/soc/renesas/r8a774a1-sysc.c
+++ b/drivers/soc/renesas/r8a774a1-sysc.c
@@ -1,9 +1,9 @@
 // SPDX-License-Identifier: GPL-2.0
 /*
- * Renesas RZ/G2M System Controller
- * Copyright (C) 2018 Renesas Electronics Corp.
+ * Renesas RZ/G2M v1.3 and RZ/G2M v3.0 System Controller
+ * Copyright (C) 2018-2020 Renesas Electronics Corp.
  *
- * Based on Renesas R-Car M3-W System Controller
+ * Based on Renesas R-Car M3-W/W+ System Controller
  * Copyright (C) 2016 Glider bvba
  */
 
@@ -14,7 +14,7 @@
 
 #include "rcar-sysc.h"
 
-static const struct rcar_sysc_area r8a774a1_areas[] __initconst = {
+static struct rcar_sysc_area r8a774a1_areas[] __initdata = {
 	{ "always-on",	    0, 0, R8A774A1_PD_ALWAYS_ON, -1, PD_ALWAYS_ON },
 	{ "ca57-scu",	0x1c0, 0, R8A774A1_PD_CA57_SCU,	R8A774A1_PD_ALWAYS_ON,
 	  PD_SCU },
@@ -39,7 +39,27 @@ static const struct rcar_sysc_area r8a774a1_areas[] __initconst = {
 	{ "3dg-b",	0x100, 1, R8A774A1_PD_3DG_B,	R8A774A1_PD_3DG_A },
 };
 
+#ifdef CONFIG_SYSC_R8A774A1
 const struct rcar_sysc_info r8a774a1_sysc_info __initconst = {
 	.areas = r8a774a1_areas,
 	.num_areas = ARRAY_SIZE(r8a774a1_areas),
 };
+#endif /* CONFIG_SYSC_R8A774A1*/
+
+#ifdef CONFIG_SYSC_R8A774A3
+static int __init r8a774a3_sysc_init(void)
+{
+	rcar_sysc_nullify(r8a774a1_areas, ARRAY_SIZE(r8a774a1_areas),
+			  R8A774A1_PD_A2VC0);
+
+	return 0;
+}
+
+const struct rcar_sysc_info r8a774a3_sysc_info __initconst = {
+	.init = r8a774a3_sysc_init,
+	.areas = r8a774a1_areas,
+	.num_areas = ARRAY_SIZE(r8a774a1_areas),
+	.extmask_offs = 0x2f8,
+	.extmask_val = BIT(0),
+};
+#endif /* CONFIG_SYSC_R8A774A3 */
diff --git a/drivers/soc/renesas/rcar-sysc.c b/drivers/soc/renesas/rcar-sysc.c
index 54ea9bb..f6dedd8 100644
--- a/drivers/soc/renesas/rcar-sysc.c
+++ b/drivers/soc/renesas/rcar-sysc.c
@@ -3,6 +3,7 @@
  *
  * Copyright (C) 2014  Magnus Damm
  * Copyright (C) 2015-2017 Glider bvba
+ * Copyright (C) 2020 Renesas Electronics Corp.
  *
  * This file is subject to the terms and conditions of the GNU General Public
  * License.  See the file "COPYING" in the main directory of this archive
@@ -289,6 +290,9 @@ static const struct of_device_id rcar_sysc_matches[] __initconst = {
 #ifdef CONFIG_SYSC_R8A774A1
 	{ .compatible = "renesas,r8a774a1-sysc", .data = &r8a774a1_sysc_info },
 #endif
+#ifdef CONFIG_SYSC_R8A774A3
+	{ .compatible = "renesas,r8a774a3-sysc", .data = &r8a774a3_sysc_info },
+#endif
 #ifdef CONFIG_SYSC_R8A774B1
 	{ .compatible = "renesas,r8a774b1-sysc", .data = &r8a774b1_sysc_info },
 #endif
diff --git a/drivers/soc/renesas/rcar-sysc.h b/drivers/soc/renesas/rcar-sysc.h
index 0864ff0..bfbba4a 100644
--- a/drivers/soc/renesas/rcar-sysc.h
+++ b/drivers/soc/renesas/rcar-sysc.h
@@ -2,6 +2,7 @@
  * Renesas R-Car System Controller
  *
  * Copyright (C) 2016 Glider bvba
+ * Copyright (C) 2020 Renesas Electronics Corp.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -56,6 +57,7 @@ extern const struct rcar_sysc_info r8a7743_sysc_info;
 extern const struct rcar_sysc_info r8a7745_sysc_info;
 extern const struct rcar_sysc_info r8a77470_sysc_info;
 extern const struct rcar_sysc_info r8a774a1_sysc_info;
+extern const struct rcar_sysc_info r8a774a3_sysc_info;
 extern const struct rcar_sysc_info r8a774b1_sysc_info;
 extern const struct rcar_sysc_info r8a774c0_sysc_info;
 extern const struct rcar_sysc_info r8a774e1_sysc_info;
-- 
2.7.4

